### PR TITLE
Optimize vertex selection by maintaining sort by cost

### PR DESCRIPTION
**Pull request by NNskelly**
Related issue: none

**Description**

Combination update (on top of SimplifyModifier1 branch recent optimizations to face reconstruction):

1. Replace the O(n)-per-removed-vertex cost lookup with an (I assume) O(n*log(n)) initial sort by cost and ongoing preservation of the sorted state during vertex removal and after recomputing costs.  Maintaining indexing on removal of a now-known index is essentially identical O(n) to the preexisting lookup of the vertex to remove.  Maintaining indexing on cost recomputation is done by localized bubbling, but given costs are unlikely to change drastically, bubbling a small number of changed vertices to preserve sort, while technically still linear time, ends up being many fewer operations than scanning the entire vertex list to find the lowest cost. (on the order of dropping the processing time for a 300k vertex mesh from 2min 5s to 1min 35s on my machine)

2. Allow an optional 3rd parameter to the modify call to specify a collapse cost at which to stop reducing even if the count has not yet been met.  This allows more tailored use of the modifier to achieve a given spatial granularity on irregularly complex meshes without knowing up front exactly what percentage reduction this will entail. (current use case: dynamically generated topographies made by extruding raster images - many, MANY vertices needed to capture every pixel of the raster, but for all I know, the input may have large regions that are completely flat)

Apologies for combining these two and rushing this through- I just learned in the past hour that the project for which I've been polishing this module is cutting the entire visualization that needs it, so I probably won't have leeway to do much more revision after today.

### CODE REVIEW

File: examples/js/modifiers/SimplifyModifier.js [hunk 1]
```diff
@@ -94,25 +94,58 @@
 
 			}
 
+			// sort & reindex now.
+			// maintaining order by edge collapse cost will be cheaper
+			// than searching for lowest every time.
+			// Lowest cost at tail will minimize reindexing on removal.
+			vertices.sort((a,b)=>{return b.collapseCost-a.collapseCost;});
```
Comment (mrdoob) [inline examples/js/modifiers/SimplifyModifier.js:101]: You can use this tool to format the code:
https://zz85.github.io/mrdoobapproves/

**Comment by NNskelly** [comment]
Done, and fixed 2 lines in the recently accepted PR that were nonconformant.

File: examples/jsm/modifiers/SimplifyModifier.js [hunk 2]
```diff
@@ -109,22 +109,63 @@ class SimplifyModifier {
...
+
 		let nextVertex;
 
 		let z = count;
 
-		while ( z -- ) {
+		// likely only a minor optimization,
+		// but why check for threshold every iteration
+		// in cases when we aren't even using it?
+		if ( threshold ) {
```
Comment (mrdoob) [inline examples/jsm/modifiers/SimplifyModifier.js:134]: Is `threshold` a boolean?

Comment (NNskelly) [reply examples/jsm/modifiers/SimplifyModifier.js:134]: Good question- no it is not.  If provided, it is a numeric value in whatever units collapseCost is in, but it is intended to be optional so as not to break existing usage.  A value of 0 would, moreover, result in no actual vertex removal (the first vertex found would already exceed the threshold), in which case the same behavior could be achieved by setting count to 0, and in my initial conception, this was consequently a safe case to ignore.   If `typeof( threshold ) == 'number' && threshold >= 0` is a more consistent, efficient, or readable-per-unit-efficiency check than relying on js truthy resolution of numeric values, that's a quick change.

Comment (mrdoob) [reply examples/jsm/modifiers/SimplifyModifier.js:134]: Ah, I think `if ( threshold >= 0 )` should do the trick then.

**Comment by NNskelly** [comment]
not sure what shifted in the last week, but auto-merge apparently couldn't deal with a single rogue newline?  Anyway, hand-merged along with a few more lines of comment that look like they didn't make last commit.

**Comment by Mugen87** [comment]
Sorry for the late review!

> Apologies for combining these two and rushing this through- I just learned in the past hour that the project for which I've been polishing this module is cutting the entire visualization that needs it, so I probably won't have leeway to do much more revision after today.

Understandable, but still it makes it harder to review the PR. Especially since the first change of the PR (the sort) is more likely to get merged than the additional `threshold` parameter (which seems too use case specific to me, tbh).

Ideally, we start with a fresh PR which only implements the sort. We can then verify how the improved time complexity of the new code actually improves the overall performance (especially with large models).

**Pull request closed without merging**